llvm/lib/CheerpUtils/PartialExecuter.cpp

```diff
@@ -0,0 +1,1777 @@
+		{
+			if (!GVar->hasInitializer() || GVar->isExternallyInitialized())
+				return false;
+		//llvm::errs() << "globalVar" << "\n";
+		}
+		//TODO: actually in opt we got some problems since we might have to find, given a ConstantExpression, wether it's internal or external
+		if (const ConstantExpr* CE = dyn_cast<ConstantExpr>(V))
+		{
+	//		for (auto op : CE->operands())
+	//		{
+	//			if (!isValueComputedConstant(op))
+	//				return false;
+	//		}
+//		llvm::errs() << *V << "\n";
+//			llvm::errs() << "is CE\n";
+		}
+		return true;
+	}
+	return false;
+}
+
+typedef std::pair<const llvm::Value*, GenericValue> ValueGenericValuePair;
+typedef std::vector<ValueGenericValuePair> LocalState;
+typedef std::unordered_map<const llvm::Value*, GenericValue> LocalStateMAP;
+
+
+
+class PartialInterpreter : public llvm::Interpreter {
+	const llvm::BasicBlock* fromBB{nullptr};	
+	enum BitMask : uint32_t
+	{
+		NONE = 0,
+		ALIGNED2 = 0x00000001,
+		ALIGNED4 = 0x00000003,
+		ALIGNED8 = 0x00000007,
+		ALL = 0xffffffff,
+	};
+	std::vector<std::pair<const llvm::Value*, std::pair<GenericValue, BitMask> > > incomings;
+public:
+	static void DO_STUFF(Function* pippo, CallBase* CB, CallBase* parent, PartialInterpreter& exe);
+	std::unordered_map<const llvm::Value*, BitMask> stronglyKnownBits;
+	BitMask computeStronglyKnownBits(const llvm::Instruction& I)
+	{
+		BitMask min = BitMask::ALL;
+
+		for (auto& op : I.operands())
+		{
+			min = BitMask(min & getBitMask(op));
+		}
+
+		//TODO: for phis just copy them
+		//TODO: also SelectInst can be specialcased
+		if (min == BitMask::ALL)
+		{
+			//All operands fully known
+			return BitMask::ALL;
+		}
+
+		switch (I.getOpcode())
+		{
+			case Instruction::Load:
+			{
+				//TODO: actually any check is needed??	YES
+				return BitMask::ALL;
+			}
+			case Instruction::Add:
+			case Instruction::Sub:
+			case Instruction::Mul:
+			case Instruction::PtrToInt:
+			case Instruction::BitCast:
+			case Instruction::GetElementPtr:
+			{
+				//Those all work modulo N
+				return min;
+			}
+			case Instruction::And:
+			{
+				//If either is Constant && less than equal min -> ALL
+				if (const ConstantInt* CI = dyn_cast<ConstantInt>(I.getOperand(0)))
+				{
+					if (CI->getZExtValue() <= min)
+						return BitMask::ALL;
+				}
+				if (const ConstantInt* CI = dyn_cast<ConstantInt>(I.getOperand(1)))
+				{
+					if (CI->getZExtValue() <= min)
+						return BitMask::ALL;
+				}
+
+				[[clang::fallthrough]];
+			}
+			case Instruction::Or:
+			case Instruction::Xor:
+			{
+				//Those all work modulo 2^k
+				return min;
+			}
+			default:
+			{
+				//Some operands NOT fully known, means in general no information is conserved
+				return BitMask::NONE;
+			}
+		}
+	}
+	std::unordered_set<const llvm::Value*> computed;
+	void setIncomingBB(const llvm::BasicBlock* from)
+	{
+		fromBB = from;
+	}
+	const llvm::BasicBlock* getIncomingBB() const
+	{
+		return fromBB;
+	}
+	enum VisitingPolicy { NORMAL, REMOVE_VALUES};
```

Comment:
The REMOVE_VALUES policy can be most likely removed, it seems to be used only when marking everything as reachable, but there are probably simpler approaches, for example we could keep the computed map as a stack in parallel with SCCs